### PR TITLE
Use recovery keys over passphrases

### PR DESCRIPTION
**Pull request by dbkr**
Give new users (or any newly set-up encryption) recovery keys and only recovery keys, and support passphrases as a secondary option for users whose accounts already have them.

This is the first step in a larger process of splitting up cross-signing and SSSS setup, but this is just the initial work to switch the priority of recovery keys and passphrases at the UI level.

Things not included in this PR:
 * Key upload button (https://github.com/vector-im/riot-web/issues/13766)
 * New logo for recovery key screens (https://github.com/vector-im/riot-web/issues/13892)
 * Centre justification of header in 'Create Recovery Key' screen (https://github.com/vector-im/riot-web/issues/13893)

I've split these out into separate issues, as indicated

How it looks currently:

After registering / logging into an account with no cross-signing:
Screenshot 2020-06-02 at 18 12 43
Screenshot 2020-06-02 at 18 12 52
Screenshot 2020-06-02 at 18 12 59

Logging in to an account with key backup:
Screenshot 2020-06-02 at 18 14 14

Logging in to an account with cross-signing (note added reset link):
Screenshot 2020-06-02 at 18 15 05

On clicking reset link:
Screenshot 2020-06-02 at 18 15 34

On clicking 'use recovery key':
Screenshot 2020-06-02 at 18 16 08

Screenshot 2020-06-02 at 18 17 00
Screenshot 2020-06-02 at 18 17 25
Screenshot 2020-06-02 at 18 18 40

Apologies for the sizeable PR: the alternative was landing it in chunks and having develop be in a weird, half-passphrase, half-recovery key state.

Fixes https://github.com/vector-im/riot-web/issues/13825

### CODE REVIEW

File: res/css/structures/auth/_CompleteSecurity.scss [hunk 1]
```diff
@@ -70,6 +70,10 @@ limitations under the License.
     }
 }
 
+.mx_CompleteSecurity_recoveryKeyInput {
+    width: 368px;
```
Comment (turt2live) [inline res/css/structures/auth/_CompleteSecurity.scss:74]: this feels precise enough that it needs a comment to explain why 368 was chosen and not 370. Something like `// dialog width minus 20px padding on either side` (as I presume this is essentially what it is?)

File: src/async-components/views/dialogs/secretstorage/CreateSecretStorageDialog.js [hunk 2]
```diff
@@ -481,194 +428,62 @@ export default class CreateSecretStorageDialog extends React.PureComponent {
...
-                    labelAllowedButUnsafe={_td("Great! This recovery passphrase looks strong enough.")}
-                />
-            </div>
-
-            <LabelledToggleSwitch
-                label={ _t("Back up encrypted message keys")}
-                onChange={this._onUseKeyBackupChange} value={this.state.useKeyBackup}
-            />
+        const DialogButtons = sdk.getComponent('views.elements.DialogButtons');
+        const InlineSpinner = sdk.getComponent("elements.InlineSpinner");
```
Comment (turt2live) [inline src/async-components/views/dialogs/secretstorage/CreateSecretStorageDialog.js:434]: Please import components

File: src/async-components/views/dialogs/secretstorage/CreateSecretStorageDialog.js [hunk 3]
```diff
@@ -481,194 +428,62 @@ export default class CreateSecretStorageDialog extends React.PureComponent {
...
     }
 
     _renderBusyPhase() {
         const Spinner = sdk.getComponent('views.elements.Spinner');
         return <div>
             <Spinner />
-        </div>;
-    }
+         </div>;
+     }
```
Comment (turt2live) [inline src/async-components/views/dialogs/secretstorage/CreateSecretStorageDialog.js:486]: indentation

**Comment by nadonomy** [comment]
@dbkr this screen doesn't make sense to me:

> Logging in to an account with key backup:
> Screenshot 2020-06-02 at 18 14 14

What's the user context exactly?

If it's logging into an account to input a key, but there are no other sessions, then I think a variation of [this screen](https://user-images.githubusercontent.com/986903/83549298-01142500-a4fd-11ea-8d29-0d8a119978b1.png) is probably what we want.

**Comment by dbkr** [comment]
Yeah, this is the user either logging in or using a session on an account that has not yet set up SSSS, has a key backup but that key backup is not trusted, so we need to establish trust in the key backup and then convert it into a SSSS store.

**Comment by dbkr** [comment]
Done. Now looks like:

Screenshot 2020-06-03 at 12 22 40
Screenshot 2020-06-03 at 12 22 48

...and then once you enter a correct passphrase or recovery key, it takes you straight into the app.

**Comment by JOJ0** [comment]
Hi,
I am not sure if this is the correct thread or if I should've better opened a new issue (but there are so many similar ones already). At least you are very much discussing and improving how texts in the cross-signing dialogues can be improved that's why I think my issue would fit well in here and does not need a new issues. Please kick me out of here if this is not the right place.

To me this dialog is misleading (with our without the newly added "reset link" you discussed above:

https://user-images.githubusercontent.com/986903/83549298-01142500-a4fd-11ea-8d29-0d8a119978b1.png

I have this feeling since this dialogue came up in 1.6.0. I tried to read it differently and understand it's true meaning again and again but still come to the conclusion that it's to be understood like the following (and this is not what you actually mean):

- I should now go to one of my other sessions and do something *over there*
- Then I come back here and this dialogue will have magically gone away
- As a last resort method I could also klick on the "Use Recovery Key" link to make this dialogue go away and finalize the verification.

I know this is not what you want from me. You want:

- Please click on the tiny green link that says "use recovery key" to achive what is written in the _headline_ of this dialogue. this is your only option
- if you forgot your passphrase just create a new one (with the new reset link)

some issues that add to the IMHO misleading texts:

- the riot web and riot mobile app "symbols" very much look like klickable buttons - i am forced to try to klick on them each time i see this dialogue ;-)
- the green link is tiny and it does not look like as being the one (and only one) "correct" thing to do now (couldn't it just look like the skip button but being green)

sorry for hijacking this thread again and i hope you understand what i mean
all the best
thanks for your great work!
Jojo

**Comment by t3chguy** [comment]
> I should now go to one of my other sessions and do something over there

This is correct. 

> Then I come back here and this dialogue will have magically gone away

It won't magically go away but you will be given a new prompt to proceed with a QR code/bunch of Emojis after completing which it'll go away.

> As a last resort method I could also klick on the "Use Recovery Key" link to make this dialogue go away and finalize the verification

Precisely.

It seems you understood it right.

**Comment by JOJ0** [comment]
Ok @t3chguy, thanks a lot for clarifying!

Then my problem is somewhere else because funilly enough I just yesterday realized that the "magic" you just described is properly happening when I use my private riot installation which is only riot-web-1.5.15 and the latest mobile ios client.

when i described above i was referring to our installation at work. there we have riot-web 1.6.2 and no mobile clients.

i could not get this dialog go away (and a colleague told me that it is supposed to "only go away" when i put in my recovery passphrase and I am misunderstanding it's meaning). that was my intention to file this issues, well kind of "hijack it". sorry again! :-)))

i understood what i described above and went to _all_ my other sessions but the verifying dialogs (qr-code, emojis, etc. as you just described) never appeared on any of them!

i also double checked my user account's "verified sessions" dialog, to find out that all my sessions seem to be verified:

FIXME screenshot

so somehow i probably went into a bug, believed my colleague too much and went down the wrong road when i hijacked this issue haha ;-)

still it would be interesting to find out what the problem actually was....not sure if i could reproduce it......

**Comment by t3chguy** [comment]
Very weird as `riot-web-1.5.15` doesn't support cross-signing at all.

**Comment by JOJ0** [comment]
wait a minute: no, i am very sorry, that was a wrong information! i have a bunch of test installations lying around: 1.6.x rc1 rc2 and so on. it must have been one of those when i tried it (to find out if it is similar to the behaviour at work!)

i can remember that there was kind of a "wheel" spinning in the dialogue - signalizing that it is waiting that i go over to my other session and do something there. sorry can't remember exactly but i think it was either 1.6.0* or 1.6.1*. you patched this spinning wheel out again in 1.6.2 or 1.6.3, right?

**Comment by JOJ0** [comment]
update: it most probably was 1.6.0 because it is still open in my browser on my laptop.

anyway back to the original topic: I updated my post above with the missing screenshot of the "verified sessions". It shows that everything is verified. Is it possible that for some reason the "new and to be verified session" never made it into this list so the other sessions didn't know about it and there was no reason to show the "verification dialogue".

well, probably my vague error descriptions don't help much to really find out what the problem was, so let's leave that be and when i find a way to reproduce it, I will open a new issue.

you helped me already by clarifiying that i understood correctly and describing the way it is supposed to work.

one last thing: the "spinning wheel" in the verification dialogue made it very clear to me that there is something to do in "another compatible (1.6.0+) session and would have probably prevented the different understandings of my and my colleauge ;-))) I am sure there were reasons why it had to go away again, I am just reporting my personal user experience here :-))

thank you very much and keep up the good work! appreciated!!!!

**Pull request merged**